### ISSUE TITLE
sys-kernel/linux-image does not support generating initramfs

### ISSUE THREAD

**Issue by PF4Public** [PF4Public/gentoo-overlay#16]
Implementing (someday…) support for generating initramfs in sys-kernel/linux-image would be nice.

**Comment by keks24** [comment]
@PF4Public 
Maybe you can use this? :) https://github.com/keks24/update-initramfs

**Comment by PF4Public** [comment]
Quite probable! Your script could be a separate ebuild, which could be a conditional build-time dependency for sys-kernel/linux-image.
I'm not using initramfs on my systems however, so I'm not planning on implementing it in an immediate future by myself, but your script looks very… passend, ehm… suitable.

**Comment by keks24** [comment]
Great idea to create an ebuild out of that!
I actually made a [gentoo overlay](https://github.com/keks24/gentoo-overlay) for the first time recently and my plan is to create an ebuild for `torbrowser-launcher-bash`.

If you are using Gentoo and your current Kernel directory points to or is `/usr/src/linux/`, you should be able to use the script `update_initramfs` out of the box. It is just preparing everything in `/usr/src/iniramfs/` and the the `compressed cpio gzip archive` will be in `/usr/src/`. The installation itself is intentionally done manually.

In my opinion `update-initramfs` is not quite finished yet: I am still looking for a decent way to implement graphics drivers, so you do not have to use the EFI driver thingy, which only can provide a resolution of 1024x768.

**Comment by PF4Public** [comment]
> Great idea to create an ebuild out of that!

And for it to be invoked from another ebuild (like sys-kernel/linux-image for instance) or by end-user, it needs to be modified a bit to become more universal.

> If you are using Gentoo and your current Kernel directory points to or is `/usr/src/linux/`, you should be able to use the script `update_initramfs` out of the box. It is just preparing everything in `/usr/src/iniramfs/` and the the `compressed cpio gzip archive` will be in `/usr/src/`. The installation itself is intentionally done manually.

Indeed, that's what I'm writing about. You have a lot of variables hardcoded. So, if I'm using you script to build an initramfs for sys-kernel/linux-image, it might not go well, as portage jails every compilation into a sandbox with limited outside interactions, so the assumption of kernel location in `/usr/src/linux` would be wrong. Furthermore, due to sandbox restrictions even if one's script does not touch anything outside sandbox, other scripts, that get called, could and most definitely would do so, breaking the compilation in the end. That is actually the reason, why [I have to copy sources into the sandbox](https://github.com/PF4Public/gentoo-overlay/blob/a500599a1df9ca1c48d0faf687faea6692104a6d/sys-kernel/linux-image/linux-image-4.14.178.ebuild#L30-L33) before compiling kernel.

> Hö. Wir könnten uns hier theoretisch auf Deutsch unterhalten und nur wenige Leute würden uns hier verstehen.

… und andere würden sich den Kopf darüber zerbrechen, worum es doch geht. :D Es wäre aber nicht so nett von uns, besonders wenn alle englisch schreiben. ;)

**Comment by keks24** [comment]
> … und andere würden sich den Kopf darüber zerbrechen, worum es doch geht. :D Es wäre aber nicht so nett von uns, besonders wenn alle englisch schreiben. ;)

Yes, let's keep it real. :)

> Indeed, that's what I'm writing about. You have a lot of variables hardcoded.

Oh yes, I know what you mean, like having the opportunity to define the compression method and so forth.

> So, if I'm using you script to build an initramfs for sys-kernel/linux-image, it might not go well, as portage jails every compilation into a sandbox with limited outside interactions, so the assumption of kernel location in /usr/src/linux would be wrong. Furthermore, due to sandbox restrictions even if one's script does not touch anything outside sandbox, other scripts, that get called, could and most definitely would do so, breaking the compilation in the end. That is actually the reason, why I have to copy sources into the sandbox before compiling kernel.

Yeah, the sandbox. But it has its purpose to not screw things up on the actual filesystem, which is good!

When I get there to create an ebuild, I will see what I have to adapt. I am still new to this topic and I am using your repository as inspiration source already. :D

Alright, I will come back sooner or later about this topic!

-keks24

**Comment by PF4Public** [comment]
> Yeah, the sandbox. But it has its purpose to not screw things up on the actual filesystem, which is good!

Exactly! You either play by the rules, or else need to implement ugly kludges in your code.

> inspiration source already

You might also find interesting (and hopefully useful) [this ebuild](https://github.com/PF4Public/gentoo-overlay/blob/master/app-admin/gb-chroot/gb-chroot-9999.ebuild) and a [corresponding source repository](https://github.com/PF4Public/gb-chroot).

**Comment by PF4Public** [comment]
Recently I discovered that with `CONFIG_INITRAMFS_SOURCE` kernel could create initramfs during the build. I suppose, this could be useful.

> The config option CONFIG_INITRAMFS_SOURCE (in General Setup in menuconfig,
> and living in usr/Kconfig) can be used to specify a source for the
> initramfs archive, which will automatically be incorporated into the
> resulting binary.  This option can point to an existing gzipped cpio
> archive, a directory containing files to be archived, or a text file
> specification such as the following example:
> 
>   dir /dev 755 0 0
>   nod /dev/console 644 0 0 c 5 1
>   nod /dev/loop0 644 0 0 b 7 0
>   dir /bin 755 1000 1000
>   slink /bin/sh busybox 777 0 0
>   file /bin/busybox initramfs/busybox 755 0 0
>   dir /proc 755 0 0
>   dir /sys 755 0 0
>   dir /mnt 755 0 0
>   file /init initramfs/init.sh 755 0 0

**Comment by keks24** [comment]
> I mean, this way kernel compilation and initramfs generation could be separated and not depended on each other.

It is actually the opposite. If you specify a `gzipped cpio archive` file, `directory containing files to be archived` or `a text file specification`, the files must exist before compiling the kernel. So you are giving it an external dependency more or less. Moreover the `initramfs` will `be incorporated into the resulting binary`, which I wanted to avoid with `update_initramfs`.

**Comment by PF4Public** [comment]
> Moreover the `initramfs` will `be incorporated into the resulting binary`, which I wanted to avoid with `update_initramfs`.

Quite reasonable. There must be a configurable option then.

**Comment by Techwolf** [comment]
sys-kernel/gentoo-kernel::gentoo does build and install the kernel with initramfs USE flag. Can use custom .config if needed.

More info at https://wiki.gentoo.org/wiki/Kernel and https://wiki.gentoo.org/wiki/Project:Distribution_Kernel

I use this as a backup kernel along with my custom config zen kernel that I use all the time.
